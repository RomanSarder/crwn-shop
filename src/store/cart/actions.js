import { createAction } from "@reduxjs/toolkit";

export var toggleCart = createAction('cart/toggle')
export var addItemToCart = createAction('cart/addItem')
export var removeItemFromCart = createAction('cart/removeItem')
export var increaseItemQuantity = createAction('cart/increaseQuantity')
export var decreaseItemQuantity = createAction('cart/decreaseQuantity')