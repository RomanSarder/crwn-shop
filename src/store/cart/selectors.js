export function selectCartDisplayState (state) {
    return state.cart.showCart
}