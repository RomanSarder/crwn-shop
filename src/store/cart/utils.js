export function getCartItemById (state, id) {
    const itemIndex = state.cartItems.findIndex(function findPendingItem (item) {
        return item.id == id
    })

    var targetItem = state.cartItems[itemIndex]

    return {
        index: itemIndex,
        item: targetItem
    }
}

export function removeCartItem (state, id) {
    return state.cartItems.filter(function filterInItemsExceptId (item) {
        return item.id !== id
    })
}