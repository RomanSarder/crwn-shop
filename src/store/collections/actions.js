import { createAction, createAsyncThunk } from "@reduxjs/toolkit";
import { getCollectionsWithItems } from "../../api/collections";

export const setCollectionsData = createAction('collections/setData')
export const collectionLoadingFinished = createAction('collections/loadingFinished')
export const collectionLoadingStarted = createAction('collections/loadingStarted')

export const getCollectionsData = createAsyncThunk(
    'collections/getData',
    async function getData (_, { dispatch }) {
        dispatch(collectionLoadingStarted())
        
        try {
            var response = await getCollectionsWithItems()
            dispatch(setCollectionsData(response))
        } finally {
            dispatch(collectionLoadingFinished())
        }
    },
    {
        condition: function checkIfDataAlreadyLoaded (_, { getState }) {
          const { collections: { data, isFetching } } = getState()
          const isDataAlreadyLoaded = Object.keys(data).length > 0
          if (isDataAlreadyLoaded || isFetching) {
            // Already in progress or data was loaded
            return false
          }
        }
    }
)