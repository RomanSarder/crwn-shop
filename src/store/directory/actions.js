import { createAsyncThunk, createAction } from "@reduxjs/toolkit";
import { getDirectoryData as getDirectoryDataFromAPI } from "../../api/directory";

export var setDirectoryData = createAction('directory/setData')
export var directoryLoadingFinished = createAction('directory/loadingFinished')
export var directoryLoadingStarted = createAction('directory/loadingStarted')

export var getDirectoryData = createAsyncThunk(
    'directory/getData',
    async function getData (_, { dispatch }) {
        dispatch(directoryLoadingStarted())
        
        try {
            var response = await getDirectoryDataFromAPI()
            dispatch(setDirectoryData(response))
        } finally {
            dispatch(directoryLoadingFinished())
        }
    },
    {
        condition: function checkIfDataAlreadyLoaded (_, { getState }) {
          var { directory: { items, isFetching } } = getState()
          const isDataAlreadyLoaded = items.length > 0
          if (isDataAlreadyLoaded || isFetching) {
            // Already in progress or data was loaded
            return false
          }
        }
    }
)