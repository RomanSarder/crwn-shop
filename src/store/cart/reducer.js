import { createReducer } from "@reduxjs/toolkit";
import { addItemToCart, removeItemFromCart, toggleCart } from "./actions";

var initialState = {
    showCart: false,
    cartItems: [],
}

export default createReducer(initialState, function buildReducer (builder) {
    builder
        .addCase(toggleCart, function updateState (state) {
            state.showCart = !state.showCart
        })
        .addCase(addItemToCart, function updateState (state, { payload }) {
            var currentCartItems = state.cartItems
            var pendingItem = { ...payload, quantity: 1 }

            const sameItemIndexInCartItems = currentCartItems.findIndex(function findPendingItem (item) {
                return item.id == pendingItem.id
            })
            const isItemAlreadyInCart = sameItemIndexInCartItems > -1

            if (isItemAlreadyInCart) {
                var targetCartItem = state.cartItems[sameItemIndexInCartItems]
                state.cartItems[sameItemIndexInCartItems] = { ...targetCartItem, quantity: targetCartItem.quantity + 1 }
            } else {
                state.cartItems.push({ ...pendingItem, quantity: 1 })
            }
        })
        .addCase(removeItemFromCart, function updateState (state, { payload }) {
            var newCartItems = state.cartItems.filter(function filterInItemsExceptId (item) {
                return item.id !== payload
            })

            state.cartItems = newCartItems
        })
})