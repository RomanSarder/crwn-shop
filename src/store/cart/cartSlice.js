import { persistReducer } from 'redux-persist'
import { cartPersistConfig } from '../persist-config'
import reducer from './reducer'

export * as actions from './actions'
export * as selectors from './selectors'
export default persistReducer(cartPersistConfig, reducer)