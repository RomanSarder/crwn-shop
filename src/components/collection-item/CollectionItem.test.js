import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import 'jest-styled-components'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { addItemToCart } from '../../store/cart/actions'
import CollectionItemContainer, { CollectionItem } from './CollectionItem'

var testItem = {
    id: 1,
    price: 10,
    name: 'test item 1',
    imageUrl: 'test image url 1'
}

function renderCollectionItemComponent ({ onAddToCart = jest.fn() } = {}) {
    var renderResult = render(<CollectionItem {...testItem} onAddToCart={onAddToCart}/>)

    return renderResult
}

function renderCollectionItemContainer (store) {
    var renderResult = render(
        <Provider store={store}>
            <CollectionItemContainer {...testItem} />
        </Provider>
    )

    return renderResult
}

it('<CollectionItem /> should properly render', () => {
    var { container, getByText, getByRole } = renderCollectionItemComponent()

    expect(container.firstChild).toMatchSnapshot()
    expect(getByText(testItem.name)).toBeVisible()
    expect(getByText(`$${testItem.price}`)).toBeVisible()
    expect(getByRole('button', { hidden: true })).toHaveTextContent('Add To Cart')
})

it('<CollectionItem /> should properly call handle when Add To Cart clicked', () => {
    var onAddToCartMock = jest.fn()
    var { getByRole } = renderCollectionItemComponent({ onAddToCart: onAddToCartMock })

    userEvent.click(getByRole('button', { hidden: true }))

    expect(onAddToCartMock).toHaveBeenCalledTimes(1)
})

it('should properly dispatch addToCart action on button click', () => {
    var store = configureStore()()
    var { getByRole } = renderCollectionItemContainer(store)

    userEvent.click(getByRole('button', { hidden: true }))

    expect(store.getActions()[0]).toEqual({
        type: addItemToCart().type,
        payload: testItem
    })
})