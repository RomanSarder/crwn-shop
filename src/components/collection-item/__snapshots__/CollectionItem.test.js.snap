// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollectionItem /> should properly render 1`] = `
.c1 {
  min-width: 16.5rem;
  width: auto;
  height: 5rem;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 5rem;
  padding: 0 3.5rem 0 3.5rem;
  font-size: 1.5rem;
  background-color: black;
  color: white;
  text-transform: uppercase;
  font-family: 'Open Sans Condensed';
  font-weight: bolder;
  cursor: pointer;
  border: none;
  background-color: white;
  color: black;
  border: 2px solid black;
}

.c1:hover {
  background-color: white;
  color: black;
  border: 1px solid black;
}

.c1:hover {
  background-color: black;
  color: white;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
}

.c0 .add-to-cart-btn {
  position: absolute;
  visibility: hidden;
  opacity: 0.7;
  width: 80%;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,200%);
  -ms-transform: translate(-50%,200%);
  transform: translate(-50%,200%);
}

.c0 .add-to-cart-btn:hover {
  opacity: 1;
}

.c0:hover .image-container {
  opacity: 0.8;
}

.c0:hover .add-to-cart-btn {
  visibility: visible;
}

.c0 .image-container {
  background-image: url(test image url 1);
  background-size: cover;
  background-position: center;
  height: 38rem;
  width: 34rem;
}

.c0 .info {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 .info span {
  font-size: 2.2rem;
}

<div
  class="c0"
>
  <div
    class="image-container"
  />
  <div
    class="info"
  >
    <span>
      test item 1
    </span>
    <span>
      $
      10
    </span>
  </div>
  <button
    class="c1 add-to-cart-btn"
  >
    Add To Cart
  </button>
</div>
`;
